src/RestApi/StoreApi/Routes/Checkout.php

```diff
@@ -43,28 +45,289 @@ public function get_args() {
 			[
 				'methods'  => \WP_REST_Server::CREATABLE,
 				'callback' => [ $this, 'get_response' ],
-				// @todo Determine the args we want to accept here.
-				'args'     => [],
+				'args'     => [
+					'order_id'       => [
+						'description' => __( 'The order ID being processed.', 'woo-gutenberg-products-block' ),
+						'type'        => 'number',
+					],
+					'order_key'      => [
+						'description' => __( 'The order key; used to validate the order is valid.', 'woo-gutenberg-products-block' ),
+						'type'        => 'string',
+					],
+					'payment_method' => [
+						'description' => __( 'The ID of the payment method being used to process the payment.', 'woo-gutenberg-products-block' ),
+						'type'        => 'string',
+					],
+					'payment_data'   => [
+						'description' => __( 'Data needed to take payment via the chosen payment method. This is passed through to the gateway when processing the payment for the order.', 'woo-gutenberg-products-block' ),
+						'type'        => 'array',
+						'items'       => [
+							'type'       => 'object',
+							'properties' => [
+								'key'   => [
+									'type' => 'string',
+								],
+								'value' => [
+									'type' => 'string',
+								],
+							],
+						],
+					],
+				],
 			],
 			'schema' => [ $this->schema, 'get_public_item_schema' ],
 		];
 	}
 
 	/**
-	 * Convert the cart into a new draft order, or update an existing draft order, and return an updated cart response.
+	 * Process a given order and generate a response for the endpoint.
 	 *
 	 * @throws RouteException On error.
 	 * @param \WP_REST_Request $request Request object.
 	 * @return \WP_REST_Response
 	 */
 	protected function get_route_post_response( \WP_REST_Request $request ) {
-		// @todo Add order and payment gateway processing here.
-		// @todo we need to determine the fields we want to return after processing e.g. redirect URLs.
-		$checkout_result = [
-			'order' => new \WC_Order(),
-		];
-		$response        = $this->prepare_item_for_response( $checkout_result, $request );
-		$response->set_status( 200 );
+		$order = $this->get_request_order_object( $request );
+
+		$this->update_order_before_payment( $order, $request );
+
+		if ( ! $order->needs_payment() ) {
+			$payment_result = $this->process_without_payment( $order, $request );
+		} else {
+			$payment_result = $this->process_payment( $order, $request );
+		}
+
+		$response = $this->prepare_item_for_response(
+			[
+				'order_id'       => $order->get_id(),
+				'payment_result' => $payment_result,
+			],
+			$request
+		);
+
+		switch ( $payment_result->status ) {
+			case 'success':
+				$response->set_status( 200 );
+				break;
+			case 'pending':
+				$response->set_status( 202 );
+				break;
+			case 'failure':
+				$response->set_status( 400 );
+				break;
+			case 'error':
+				$response->set_status( 500 );
+				break;
+		}
+
```

Comment:
It appears this is the replacement for the `WC_Checkout::process_checkout`  behaviour right? Should we make note somewhere (maybe in a followup issue?) of all the things that are _not_ running through this logic that _are_ running with the logic on `WC_Checkout::process_checkout`? For instance, from a brief look this is a list I came up with:

- `wc_set_time_limit`,
- `woocommerce_before_checkout_process` action hook
- handling if cart is empty (session expired?)
- `woocommerce_checkout_process` action hook
- `WC_Checkout::validate_checkout` it looks like we do the same logic in this endpoint, but just checking.
- `woocommerce_checkout_order_processed` action

Reply:
Yes we need a follow up to discuss the hooks in particular. I am concerned that throwing in hooks may further complicate things under this context.